pkg/resourceinterpreter/customized/webhook/serviceresolvers.go

```diff
@@ -0,0 +1,80 @@
+package webhook
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	webhookutil "k8s.io/apiserver/pkg/util/webhook"
+	corev1 "k8s.io/client-go/listers/core/v1"
+)
+
+// ServiceResolver knows how to convert a service reference into an actual location.
+type ServiceResolver interface {
+	ResolveEndpoint(namespace, name string, port int32) (*url.URL, error)
+}
+
+// NewServiceResolver returns a ServiceResolver that parses service first,
+// if service not exist, constructs a service URL from a given namespace and name.
+func NewServiceResolver(services corev1.ServiceLister) ServiceResolver {
+	return &serviceResolver{
+		services:        services,
+		defaultResolver: webhookutil.NewDefaultServiceResolver(),
+	}
+}
+
+type serviceResolver struct {
+	services        corev1.ServiceLister
+	defaultResolver ServiceResolver
+}
+
+func (r *serviceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
+	res, err := resolveCluster(r.services, namespace, name, port)
+	if err != nil && apierrors.IsNotFound(err) {
+		return r.defaultResolver.ResolveEndpoint(namespace, name, port)
+	}
+	return res, err
+}
+
+// resolveCluster parses Service resource to url.
+// It is lifted from https://github.com/kubernetes/apiserver/blob/release-1.31/pkg/util/proxy/proxy.go#L105.
+func resolveCluster(services corev1.ServiceLister, namespace, id string, port int32) (*url.URL, error) {
```

Comment:
```suggestion
func resolveCluster(namespace, id string, port int32) (*url.URL, error) {
```
We can try to remove the service lister from the function signatures; this will make it easy to build a unit test.